README.md
# setriq: pairwise sequence distances
[![CircleCI](https://circleci.com/gh/BenTenmann/setriq/tree/main.svg?style=shield&circle-token=11d21cf82d1b29647f02543f6bfee9703a8f7bfe)](https://circleci.com/gh/BenTenmann/setriq/tree/main)
[![codecov](https://codecov.io/gh/BenTenmann/setriq/branch/main/graph/badge.svg?token=I8Q4RWPCMS)](https://codecov.io/gh/BenTenmann/setriq)
[![CodeFactor](https://www.codefactor.io/repository/github/bentenmann/setriq/badge)](https://www.codefactor.io/repository/github/bentenmann/setriq)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![logo](fig/logo.png)

A `Python` package written in `C++` for computing pairwise distances between (immunoglobulin) sequences.

[**Documentation**](https://bentenmann.github.io/setriq/)

## Install
This package is available on PyPI
```bash
pip install setriq
```

## Quickstart

`setriq` inherits from the `torch` philosophy of callable objects. Each `Metric` subclass is a callable upon
initialisation, taking a list of objects (usually `str`) and returning a list of `float` values.

```python
import setriq
metric = setriq.CdrDist()

sequences = [
    'CASSLKPNTEAFF',
    'CASSAHIANYGYTF',
    'CASRGATETQYF'
]
distances = metric(sequences)
```

The returned list is flat and contains `N * (N - 1) / 2` elements, i.e. the lower (or upper) triangle of the distance
matrix. To get the square form of the matrix, use `scipy.spatial.distance.squareform` on the returned distances.

## About

As the header suggests, `setriq` is a no-frills Python package for fast computation of pairwise sequence distances, with
a focus on immunoglobulins. It is a declarative framework and borrows many concepts from the popular `torch` library. It
has been optimized for parallel compute on CPU architectures.

Available distance functions:
* CDRdist
* Levenshtein
* TCRdist
* Hamming
* Jaro
* Jaro-Winkler
* Longest Common Substring
* Optimal String Alignment

These distance functions are available either through the object-based API (as seen above), which provides the CPU-based
parallelism, or the functional API in `setriq.single_dispatch`. Unlike the object-based API, the functional API does a
single comparison between two sequences for every call, i.e. it exposes the `C++` distance functions without the
parallelism wrapper. This can be useful for integration of `setriq` with other tools such as `PySpark`. For example:

```python
from pyspark.sql import SparkSession
from pyspark.sql.functions import udf
from pyspark.sql.types import DoubleType

from setriq import single_dispatch as sd

spark = SparkSession \
   .builder \
   .appName("setriq-spark") \
   .getOrCreate()

df = spark.createDataFrame([('CASSLKPNTEAFF',), ('CASSAHIANYGYTF',), ('CASRGATETQYF',)], ['sequence'])
df = df.withColumnRenamed('sequence', 'a').crossJoin(df.withColumnRenamed('sequence', 'b'))

lev_udf = udf(sd.levenshtein, returnType=DoubleType())  # single dispatch levenshtein distance
df = df.withColumn('distance', lev_udf('a', 'b'))
df.show()
```

It is important to note, that for `setriq.single_dispatch` the returned value is always a single float value.

## Requirements
A `Python` version of 3.7 or above is required, as well as a `C++` compiler equipped with OpenMP. The package has been
tested on Linux and macOS. To get the required OpenMP resources, run:

On Linux:
```bash
sudo apt install libomp-dev && sudo apt show libomp-dev
```

On macOS:
```bash
brew install libomp llvm
```


## References
   1. Dash, P., Fiore-Gartland, A.J., Hertz, T., Wang, G.C., Sharma, S., Souquette, A., Crawford, J.C., Clemens, E.B.,
      Nguyen, T.H., Kedzierska, K. and La Gruta, N.L., 2017. Quantifiable predictive features define epitope-specific T cell
      receptor repertoires. Nature, 547(7661), pp.89-93. (https://doi.org/10.1038/nature22383)
   2. Jaro, M.A., 1989. Advances in record-linkage methodology as applied to matching the 1985 census of Tampa,
      Florida. Journal of the American Statistical Association, 84(406), pp.414-420.
   3. Levenshtein, V.I., 1966, February. Binary codes capable of correcting deletions, insertions, and reversals. In
      Soviet physics doklady (Vol. 10, No. 8, pp. 707-710).
   4. python-Levenshtein (https://github.com/ztane/python-Levenshtein)
   5. Thakkar, N. and Bailey-Kellogg, C., 2019. Balancing sensitivity and specificity in distinguishing TCR groups by CDR
      sequence similarity. BMC bioinformatics, 20(1), pp.1-14. (https://doi.org/10.1186/s12859-019-2864-8)
   6. Van der Loo, M.P., 2014. The stringdist package for approximate string matching. R J., 6(1), p.111.
   7. Winkler, W.E., 1990. String comparator metrics and enhanced decision rules in the Fellegi-Sunter model of record
      linkage.


CHANGELOG.md
## [1.6.4](https://github.com/BenTenmann/setriq/compare/deploy-1.6.3...deploy-1.6.4) (2022-12-21)


### Bug Fixes

* **ci:** add pybind dep to source dist ([#26](https://github.com/BenTenmann/setriq/issues/26)) ([e64979e](https://github.com/BenTenmann/setriq/commit/e64979ea6d60e64dbfcb37cea561d47993e9df5d))

## [1.6.3](https://github.com/BenTenmann/setriq/compare/deploy-1.6.2...deploy-1.6.3) (2022-12-21)


### Bug Fixes

* **ci:** try new filters + update gh-pages dependency ([a030b27](https://github.com/BenTenmann/setriq/commit/a030b27184c77e57a7be28f17043684a28d251e2))

## [1.6.2](https://github.com/BenTenmann/setriq/compare/deploy-1.6.1...deploy-1.6.2) (2022-12-19)


### Bug Fixes

* **ci:** update triggers ([55958a2](https://github.com/BenTenmann/setriq/commit/55958a22b2b9093dacb11aa1d53e21081be69e09))

## [1.6.1](https://github.com/BenTenmann/setriq/compare/deploy-1.6.0...deploy-1.6.1) (2022-12-18)


### Bug Fixes

* **ci:** bump semrel package versions ([dc076ec](https://github.com/BenTenmann/setriq/commit/dc076eca0638c7679aa573e1f0e7cf449869392e))
* **ci:** trigger CI ([#21](https://github.com/BenTenmann/setriq/issues/21)) ([0830f4a](https://github.com/BenTenmann/setriq/commit/0830f4af538a3502fc104a4a38eb80c6aeb4dc47))
* **ci:** update docker image to correct tag ([f02bac8](https://github.com/BenTenmann/setriq/commit/f02bac80306e6dca86a776f46b5b4cb7fc36bb08))

# [1.6.0](https://github.com/BenTenmann/setriq/compare/deploy-1.5.0...deploy-1.6.0) (2022-09-10)


### Features

* **lcs:** add new metrics ([b87f77e](https://github.com/BenTenmann/setriq/commit/b87f77e0429050bbb3bda70834e9874e1715221a))

# [1.5.0](https://github.com/BenTenmann/setriq/compare/deploy-1.4.2...deploy-1.5.0) (2022-06-05)


### Features

* **distances:** add hamming, jaro and jaro-winkler ([d9f7154](https://github.com/BenTenmann/setriq/commit/d9f7154fac2787ff80300be8c840c8ba9e5f989b))
* add landau vishkin algo ([87783fe](https://github.com/BenTenmann/setriq/commit/87783fe0044fab9c2b7142277aa464294912f89b))

## [1.4.2](https://github.com/BenTenmann/setriq/compare/deploy-1.4.1...deploy-1.4.2) (2022-01-29)


### Bug Fixes

* setup.py refactor ([663ad3d](https://github.com/BenTenmann/setriq/commit/663ad3da308ff32f8a38f67fee71fed647961e02))

## [1.4.1](https://github.com/BenTenmann/setriq/compare/deploy-1.4.0...deploy-1.4.1) (2022-01-29)


### Bug Fixes

* implicit conversion ([9720c4e](https://github.com/BenTenmann/setriq/commit/9720c4e9891a36947c469dd02e2e906bf0404339))
* implicit conversion ([b965d3f](https://github.com/BenTenmann/setriq/commit/b965d3f790fdc3ab4375bd6a0d40334fc92adba0))
* major refactor ([6735f13](https://github.com/BenTenmann/setriq/commit/6735f13b29bf23b0a0d264967718be3f98f3df7f))
* pass metric by value ([ba8f7b6](https://github.com/BenTenmann/setriq/commit/ba8f7b6817de73d5862c401ffdd16b240a27ff84))
* update string indexing ([145430b](https://github.com/BenTenmann/setriq/commit/145430bfb3ab62e53f1b54ddcb4b2fe7513ca3da))

# [1.4.0](https://github.com/BenTenmann/setriq/compare/deploy-1.3.1...deploy-1.4.0) (2022-01-15)


### Features

* add affine option to sw ([d96d1fd](https://github.com/BenTenmann/setriq/commit/d96d1fd1dbfbb1b7c5efa37603924334bcacc06d))

## [1.3.1](https://github.com/BenTenmann/setriq/compare/deploy-1.3.0...deploy-1.3.1) (2022-01-08)


### Bug Fixes

* pin setuptools ([d576789](https://github.com/BenTenmann/setriq/commit/d576789fc0469da93b3a4d54b25b73d030c10e83))

# [1.3.0](https://github.com/BenTenmann/setriq/compare/deploy-1.2.1...deploy-1.3.0) (2021-12-18)


### Features

* update cdr-dist to have thakkar default ([0e312f6](https://github.com/BenTenmann/setriq/commit/0e312f680e5e8645f8950fc40479501c8df3bd6e))

## [1.2.1](https://github.com/BenTenmann/setriq/compare/deploy-1.2.0...deploy-1.2.1) (2021-12-18)


### Bug Fixes

* setup flags ([daca55f](https://github.com/BenTenmann/setriq/commit/daca55f37438f9485695eaa843fe24f686c47b07))

# [1.2.0](https://github.com/BenTenmann/setriq/compare/deploy-1.1.2...deploy-1.2.0) (2021-12-17)


### Features

* pypi release ([24549a4](https://github.com/BenTenmann/setriq/commit/24549a4ee3614b58fadc18b3cbe734f9300ac32a))

## [1.1.2](https://github.com/BenTenmann/setriq/compare/deploy-1.1.1...deploy-1.1.2) (2021-12-17)


### Bug Fixes

* update build flag selection ([a62b482](https://github.com/BenTenmann/setriq/commit/a62b482c96acc79a41085594612f0058c0a2f7b0))

## [1.1.1](https://github.com/BenTenmann/setriq/compare/deploy-1.1.0...deploy-1.1.1) (2021-12-17)


### Bug Fixes

* sdist include ([dd7bc20](https://github.com/BenTenmann/setriq/commit/dd7bc2014c61e5d005232743c956be0c2c44accd))

# [1.1.0](https://github.com/BenTenmann/setriq/compare/deploy-1.0.2...deploy-1.1.0) (2021-12-16)


### Features

* update README ([530ba17](https://github.com/BenTenmann/setriq/commit/530ba171ff4ed23799151c11f40a44b3fac6a8f9))

## [1.0.2](https://github.com/BenTenmann/setriq/compare/deploy-1.0.1...deploy-1.0.2) (2021-12-16)


### Bug Fixes

* update cci config ([df6030a](https://github.com/BenTenmann/setriq/commit/df6030a400d9f93dde9854ed52dec2aedad72667))

## [1.0.1](https://github.com/BenTenmann/setriq/compare/deploy-1.0.0...deploy-1.0.1) (2021-12-16)


### Bug Fixes

* give releaseImage.sh execution permission ([6aded34](https://github.com/BenTenmann/setriq/commit/6aded3440baad07faa2467e8b52203803f7391c1))

# 1.0.0 (2021-12-16)


### Bug Fixes

* add --user to pip install ([1afecef](https://github.com/BenTenmann/setriq/commit/1afecefe6b9b485805eaeba385a42f28532bd697))
* add -e ([6f72f8c](https://github.com/BenTenmann/setriq/commit/6f72f8c6f1cb8f8e7e0d4451045cf2dbe9b940c3))
* add cstring header ([fddb517](https://github.com/BenTenmann/setriq/commit/fddb517050489ab3632b94a22b0514a93f75a597))
* add pip to install ([927c05f](https://github.com/BenTenmann/setriq/commit/927c05fad3a9770c493b3dbe3d1b0e2efdb67387))
* add pybind11 dependency to release job ([6d2a58d](https://github.com/BenTenmann/setriq/commit/6d2a58dbe7dd1d988c998c8d077e0f728199eadd))

src/setriq/__init__.py
"""
setriq
======

Provides:
    1. fast computation of pairwise (immunoglobulin) sequence distances
    2. parsed BLOSUM substitution matrices

About
-----
This package is a one-trick-pony which does its trick pretty well: efficient computation of pairwise sequence distances
on CPU. There are a number of distance functions available [1]_ [2]_ [3]_ [4]_, implemented in C++ and surfaced in
Python.

Examples
--------
>>> import setriq
>>> metric = setriq.CdrDist()
>>>
>>> sequences = [
...     'CASSLKPNTEAFF',
...     'CASSAHIANYGYTF',
...     'CASRGATETQYF'
... ]
>>> distances = metric(sequences)

References
----------
.. [1] Dash, P., Fiore-Gartland, A.J., Hertz, T., Wang, G.C., Sharma, S., Souquette, A., Crawford, J.C., Clemens, E.B.,
   Nguyen, T.H., Kedzierska, K. and La Gruta, N.L., 2017. Quantifiable predictive features define epitope-specific
   T cell receptor repertoires. Nature, 547(7661), pp.89-93. (https://doi.org/10.1038/nature22383)

.. [2] Levenshtein, V.I., 1966, February. Binary codes capable of correcting deletions, insertions, and reversals. In
   Soviet physics doklady (Vol. 10, No. 8, pp. 707-710).

.. [3] python-Levenshtein (https://github.com/ztane/python-Levenshtein)

.. [4] Thakkar, N. and Bailey-Kellogg, C., 2019. Balancing sensitivity and specificity in distinguishing TCR groups by
   CDR sequence similarity. BMC bioinformatics, 20(1), pp.1-14. (https://doi.org/10.1186/s12859-019-2864-8)

"""

from .modules import (
    BLOSUM45,
    BLOSUM62,
    BLOSUM90,
    CdrDist,
    Hamming,
    Jaro,
    JaroWinkler,
    Levenshtein,
    LongestCommonSubstring,
    OptimalStringAlignment,
    SubstitutionMatrix,
    TcrDist,
    single_dispatch,
)

__all__ = [
    "BLOSUM45",
    "BLOSUM62",
    "BLOSUM90",
    "CdrDist",
    "Hamming",
    "Jaro",
    "JaroWinkler",
    "Levenshtein",
    "LongestCommonSubstring",
    "OptimalStringAlignment",
    "SubstitutionMatrix",
    "TcrDist",
    "single_dispatch",
]
